WIP: LSP review

norcalli opened the pull request:
This is just used for someone to see the diff if they are curious. the final PR will be against the main repo.

JelteF committed:
termdebug.vim: reset evalFromBalloonExprResult #11309

The value is used again in case of a pointer and it will cause errors then.

Commit:
Checkpoint.

Commit:
Fix markdown type. Trip stuff.

Commit:
Clean up uri encoding stuff.

Commit:
Remove message.lua and util.lua

Integrate them into their callsites.

Commit:
Finish client rework.

h-michael commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -1,115 +0,0 @@
-  )
-end
-
-util.get_HoverContents_type = function(contents)
```
I don't care about util namespace, but I intentionally made this function public.
Because lsp has many union types to maintain backward compatibility, I wanted to avoid having users implement each of these judgments.
I thought that update_tagstack and is_CompletionList would be reused for users (or third parties) to handle lsp.

h-michael commented on runtime/lua/vim/lsp.lua:
```diff
@@ -200,16 +205,18 @@ lsp.client_has_started = function(filetype, server_name)
-      for _, client in pairs(filetype_clients) do
-        if client:is_running() then return true end
+      for _, client in ipairs(filetype_clients) do
+				if client:is_running() then
```
nit: neovim project use space instead of tab.

Commit:
Wooh. final refactoring.

rpc is now detached from lsp logic.

callback is likely unnecessary.

Commit:
Bugfixes

Commit:
The didChange is currently broken, so send the full buffer for now.

Commit:
Clean up encoding validation.

Commit:
Clean up and handle errors.

Commit:
Remove callbacks code.

Commit:
Clean up comments.

Commit:
Start capability resolution rewrite

Add vim.tbl_deep_merge to help

TODO: use the resolved_capabilities in the handler initializations to
conditionally send things to the server.

Commit:
Merge remote-tracking branch 'origin/master' into lsp-review

Commit:
Change logger to allow lazy eval.

Commit:
Start using resolved_capabilities.

Commit:
Improve capabilities detection

Commit:
Slight cleanup

Commit:
Switch to spaces instead of tabs.

Commit:
Switch to spaces instead of tabs pt 2

Commit:
Further cleanup.

Add env and cwd options.
Improve documentation.
Improve log statements.

Commit:
Clean up protocol.

Commit:
Bugfix and enable signatureHelp

Commit:
Clean up unsupported_method handling.

Commit:
Final cleanup.

Commit:
Last cleanup.

Commit:
Remove tests for now.

Commit:
Remove out of date docs.

Commit:
Improve tests and comments.

Commit:
Oopsie

Commit:
Bugfix and improve docs and change conf signature.

Commit:
Refactor and bugfix.

- Change callbacks to have a uniform signature.
- Clean up validation.
- Clean up scheduling.

h-michael commented:
Since PR was created upstream, this PR is no longer needed.
